metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/sync/BlockSynchronizer.scala

```diff
@@ -0,0 +1,210 @@
+  *
+  * It will walk backwards, asking for the ancestors until we find one that we already
+  * have in persistent storage, then append blocks to the storage in the opposite order.
+  *
+  * Since the final block has a Quorum Certificate, there's no need to validate the
+  * ancestors, assuming an honest majority in the federation. The only validation we
+  * need to do is hash checks to make sure we're getting the correct blocks.
+  *
+  * The synchronizer keeps the tentative blocks in memory until they can be connected
+  * to the persistent storage. We assume that we never have to download the block history
+  * back until genesis, but rather that the application will always have support for
+  * syncing to any given block and its associated state, to catch up after spending
+  * a long time offline. Once that happens the block history should be pruneable.
+  */
+class BlockSynchronizer[F[_]: Sync: Timer, N, A <: Agreement: Block](
+    blockStorage: BlockStorage[N, A],
+    getBlock: BlockSynchronizer.GetBlock[F, A],
+    inMemoryStore: KVStoreRunner[F, N],
+    semaphore: Semaphore[F],
+    retryTimeout: FiniteDuration = 5.seconds
+)(implicit storeRunner: KVStoreRunner[F, N]) {
+
+  // We must take care not to insert blocks into storage and risk losing
+  // the pointer to them in a restart, hence keeping the unfinished tree
+  // in memory until we find a parent we do have in storage, then
+  // insert them in the opposite order.
+
+  /** Download all blocks up to the one included in the Quorum Certificate.
+    *
+    * Only expected to be called once per sender at the same time, otherwise
+    * it may request the same ancestor block multiple times concurrently.
+    *
+    * This could be managed with internal queueing, but not having that should
+    * make it easier to cancel all calling fibers and discard the synchronizer
+    * instance and its in-memory store, do state syncing, then replace it with
+    * a fresh one.
+    */
+  def sync(
+      sender: A#PKey,
+      quorumCertificate: QuorumCertificate[A]
+  ): F[Unit] =
+    for {
+      path <- download(sender, quorumCertificate.blockHash, Nil)
+      _    <- persist(quorumCertificate.blockHash, path)
+    } yield ()
+
+  /** Download a block and all of its ancestors into the in-memory block store.
+    *
+    * Returns the path from the greatest ancestor that had to be downloaded
+    * to the originally requested block, so that we can persist them in that order.
+    *
+    * The path is maintained separately from the in-memory store in case another
+    * ongoing download would re-insert something on a path already partially removed
+    * resulting in a forest that cannot be traversed fully.
+    */
+  private def download(
+      from: A#PKey,
+      blockHash: A#Hash,
+      path: List[A#Hash]
+  ): F[List[A#Hash]] = {
+    storeRunner
+      .runReadOnly {
+        blockStorage.contains(blockHash)
+      }
+      .flatMap {
+        case true =>
+          path.pure[F]
```

Comment:
Are we guaranteed that if the block exists in storage then so do its ancestors?

Reply:
They could be pruned, but we'll take care that we have the last executed and committed block, so we can extend it.